fix: infer correct shard in statefulset setup (#17124, #17016)

woehrl01 opened the pull request:
Fixes #17124 
Fixes #17016
Fixes #17183

Following fixes:
1. The env for the shard env does not exist by default. This causes that there will be only shard 0 used. In a multi shard setup all clusters assigned to other shards won't be updated. 

   This PR fixes that by correctly inferring the shard in that scenario. 
   
2. The shard number comparison is accidentally done on the pointer and not the value of the shard, this of course results in different values to be compared



<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

ishitasequeira reviewed (commented):
@woehrl01 Thanks for the PR!!

I remember we have been using `defaultShardNumberValue` as `0` for application controller deployed as a StatefulSet and `-1` as the default value for the application controller deployed as a Deployment. It was intended for the default shard value if not specified via environment variable to always be 0. 

I would test out your changes once to see if the change resloves the issue.

woehrl01 commented:
@ishitasequeira you're welcome. Yes, that's why I set the value to `0` in case there is only one replica. If I see this right, this is the only case where it makes sense to hardcore it to that value, as any other scenario will make the sharding not work.

ishitasequeira reviewed (changes requested):
Thanks @woehrl01 , I tested the fix you have and was working on a fix on my end which led to the similar changes. 

There are e2e test failures though. Can you take a look at the e2e tests once?

woehrl01 commented:
@ishitasequeira Thank you, I think those are failing randomly (they have in the past). I just triggered another run.

leoluz reviewed (commented):
Tks a lot for working on @woehrl01!

Once this PR is merged we should probably work on this: https://github.com/argoproj/argo-cd/issues/17183
cc @ishitasequeira

woehrl01 commented:
@ishitasequeira 

I probably need your help here, the e2e now just uncover some strange concurrency/deadlock issues, which I also see in our 2.10. Deployment.

without any change tests are failing differently:

Run 1:

![Bildschirmfoto 2024-02-12 um 17 31 05](https://github.com/argoproj/argo-cd/assets/2535856/385ecafc-0718-4a77-80fd-0de392a39e71)

Run 2

![Bildschirmfoto 2024-02-12 um 17 31 10](https://github.com/argoproj/argo-cd/assets/2535856/0369e709-e9a0-4634-a6fa-e60e7185da16)

jannfis commented:
application-controller crashes during tests:

```
16:47:01                controller | [mINFO[0000] Updating cluster shards                      
16:47:01                controller | [mDEBU[0000] Checking if cluster https://kubernetes.default.svc with clusterShard 0 should be processed by shard 0 
16:47:01                controller | [mINFO[0000] Start watch clusters                         
16:47:01                controller | [mE0212 16:47:01.545059   15505 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
16:47:01                controller | [mgoroutine 173 [running]:
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/runtime.logPanic({0x365f340?, 0x68c0d60})
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x85
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00012f760?})
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x6b
16:47:01                controller | [mpanic({0x365f340?, 0x68c0d60?})
16:47:01                controller | [m	/opt/hostedtoolcache/go/1.21.6/x64/src/runtime/panic.go:914 +0x21f
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/sharding.hasShardingUpdates(...)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/sharding/cache.go:152
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/sharding.(*ClusterSharding).Add(0xc00012f740, 0xc00067ef00)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/sharding/cache.go:82 +0x131
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/cache.(*liveStateCache).handleAddEvent(0xc000f649a0, 0xc00067ef00)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/cache/cache.go:730 +0x3b
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/util/db.(*db).WatchClusters(0xc0010a3200, {0x47c0670, 0xc000ffff90}, 0xc000b13fd0, 0xc000b13fe0, 0xc000b13ff0)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/util/db/cluster.go:144 +0xc2
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/cache.(*liveStateCache).Run.func1()
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/cache/cache.go:717 +0x107
16:47:01                controller | [mgithub.com/argoproj/gitops-engine/pkg/utils/kube.RetryUntilSucceed.func1()
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/github.com/argoproj/gitops-engine/pkg/utils/kube/kube.go:411 +0xfb
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.PollImmediateUntil.ConditionFunc.WithContext.func1({0x4105e5, 0xc000aa9140})
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:222 +0x13
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x47c0670?, 0xc0010b12c0?}, 0xc000169550?)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:235 +0x4c
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.poll({0x47c0670, 0xc0010b12c0}, 0x56?, 0x47c0558?, 0x6cc7880?)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:582 +0x32
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.PollImmediateUntilWithContext({0x47c0670?, 0xc0010b12c0?}, 0x3a47d20?, 0xc000231638?)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:547 +0x53
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.PollImmediateUntil(0x3a47d20?, 0xc000169670, 0xc000169660?)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:538 +0x6b
16:47:01                controller | [mgithub.com/argoproj/gitops-engine/pkg/utils/kube.RetryUntilSucceed({0x47c0670?, 0xc000ffff90?}, 0x2540be400, {0x3d6a5c9, 0xe}, {{0x47ca588?, 0xc000444ac0?}, 0xc0001697b0?}, 0xc000169720)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/github.com/argoproj/gitops-engine/pkg/utils/kube/kube.go:409 +0x116
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/cache.(*liveStateCache).Run(0xc000f649a0, {0x47c0670?, 0xc000ffff90})
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/cache/cache.go:716 +0x12a
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller.(*ApplicationController).Run.func1()
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/appcontroller.go:809 +0x2e
16:47:01                controller | [mcreated by github.com/argoproj/argo-cd/v2/controller.(*ApplicationController).Run in goroutine 110
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/appcontroller.go:809 +0x6ae
16:47:01                controller | [mpanic: runtime error: invalid memory address or nil pointer dereference [recovered]
16:47:01                controller | [m	panic: runtime error: invalid memory address or nil pointer dereference
16:47:01                controller | [m[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2162311]
16:47:01                controller | [mgoroutine 173 [running]:
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00012f760?})
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xcd
16:47:01                controller | [mpanic({0x365f340?, 0x68c0d60?})
16:47:01                controller | [m	/opt/hostedtoolcache/go/1.21.6/x64/src/runtime/panic.go:914 +0x21f
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/sharding.hasShardingUpdates(...)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/sharding/cache.go:152
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/sharding.(*ClusterSharding).Add(0xc00012f740, 0xc00067ef00)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/sharding/cache.go:82 +0x131
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/cache.(*liveStateCache).handleAddEvent(0xc000f649a0, 0xc00067ef00)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/cache/cache.go:730 +0x3b
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/util/db.(*db).WatchClusters(0xc0010a3200, {0x47c0670, 0xc000ffff90}, 0xc000b13fd0, 0xc000b13fe0, 0xc000b13ff0)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/util/db/cluster.go:144 +0xc2
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/cache.(*liveStateCache).Run.func1()
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/cache/cache.go:717 +0x107
16:47:01                controller | [mgithub.com/argoproj/gitops-engine/pkg/utils/kube.RetryUntilSucceed.func1()
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/github.com/argoproj/gitops-engine/pkg/utils/kube/kube.go:411 +0xfb
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.PollImmediateUntil.ConditionFunc.WithContext.func1({0x4105e5, 0xc000aa9140})
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:222 +0x13
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x47c0670?, 0xc0010b12c0?}, 0xc000169550?)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:235 +0x4c
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.poll({0x47c0670, 0xc0010b12c0}, 0x56?, 0x47c0558?, 0x6cc7880?)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:582 +0x32
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.PollImmediateUntilWithContext({0x47c0670?, 0xc0010b12c0?}, 0x3a47d20?, 0xc000231638?)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:547 +0x53
16:47:01                controller | [mk8s.io/apimachinery/pkg/util/wait.PollImmediateUntil(0x3a47d20?, 0xc000169670, 0xc000169660?)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:538 +0x6b
16:47:01                controller | [mgithub.com/argoproj/gitops-engine/pkg/utils/kube.RetryUntilSucceed({0x47c0670?, 0xc000ffff90?}, 0x2540be400, {0x3d6a5c9, 0xe}, {{0x47ca588?, 0xc000444ac0?}, 0xc0001697b0?}, 0xc000169720)
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/vendor/github.com/argoproj/gitops-engine/pkg/utils/kube/kube.go:409 +0x116
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller/cache.(*liveStateCache).Run(0xc000f649a0, {0x47c0670?, 0xc000ffff90})
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/cache/cache.go:716 +0x12a
16:47:01                controller | [mgithub.com/argoproj/argo-cd/v2/controller.(*ApplicationController).Run.func1()
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/appcontroller.go:809 +0x2e
16:47:01                controller | [mcreated by github.com/argoproj/argo-cd/v2/controller.(*ApplicationController).Run in goroutine 110
16:47:01                controller | [m	/home/runner/work/argo-cd/argo-cd/controller/appcontroller.go:809 +0x6ae
16:47:01                controller | [mTerminating controller```

ishitasequeira commented:
Thanks @jannfis for sharing the logs. The crash seems to be because of the change `return int64(*old.Shard) != int64(*new.Shard)` part of this PR.

woehrl01 commented:
@ishitasequeira oh that's awkward 😅🙈 thank you!

Should be nil == nil handled as the same? I'm not sure in which case this makes sense.

leoluz reviewed (changes requested):
Please check my comment

leoluz commented:
> Should be nil == nil handled as the same? I'm not sure in which case this makes sense.

@woehrl01 This should return false as per the code implemented at line 149

woehrl01 commented:
@leoluz if that's the case then I would say that my check is correct.

If both are nil, it returns false. If one of them is nil, the other can't, because of the previous check. so they are different and this should return true.

jannfis commented on controller/sharding/cache.go:
```diff
@@ -149,11 +149,12 @@ func hasShardingUpdates(old, new *v1alpha1.Cluster) bool {
 		return false
 	}
-	return old.Shard != new.Shard
+	return old.Shard == nil || new.Shard == nil || int64(*old.Shard) != int64(*new.Shard)
```
Agree, all this should be backed up my unit tests.

leoluz replied:
@ishitasequeira if this is the case:
>@leoluz it should return true in that case which indicates that the clusterShard has been updated.
https://github.com/argoproj/argo-cd/pull/17167#issuecomment-1939569340

Then his code seems correct

ishitasequeira replied:
Confirmed it by testing. The current logic in the PR works.

leoluz commented:
> If both are nil, it returns false. If one of them is nil, the other can't, because of the previous check. so they are different and this should return true.
If your check should be used, I would modify the previous one to replace the && with || . Right?

@woehrl01 The case that we are trying to address is
```
new not nil
old not nil
new.Shard not nil
old.Shard is nil
```
In line 149 we have the first pass mainly making `nil` == `nil` return false. But this is not our case so we proceed to line 152.

The way it is currently implemented we have:
`old.Shard == nil || new.Shard == nil || int64(*old.Shard) != int64(*new.Shard)`

as `old.Shard` is nil the code will return `true` with is not what we want.

Thats why the proposal is to have instead:

`old.Shard != nil && new.Shard != nil && int64(*old.Shard) != int64(*new.Shard)`

where `old.Shard != nil && new.Shard != nil` is mainly doing nil checks and avoiding the nil pointer exception.

Makes sense?

woehrl01 commented:
@leoluz I'm not sure what I miss tbh. According to the algorithm we have to return true on nil shard, as it will handle that case here: https://github.com/argoproj/argo-cd/pull/17167/files#diff-f5220d702bcabf138f16d7b8e08375d96145a6073941f07df3a5801264ad5442R127

If we return true only if both are not nil, it will never be updated. Right?

leoluz commented:
> If we return true only if both are not nil, it will never be inferred and updated. Right?

@woehrl01 ok, maybe I am not understanding what this function should behave.

@ishitasequeira If we have the scenario below the function should return `true` or `false`?

```
new not nil
old not nil
new.Shard not nil
old.Shard is nil
```

ishitasequeira commented:
@leoluz it should return true in that case which indicates that the clusterShard has been updated.

woehrl01 commented:
@leoluz me neither, I just assume it from the surrounding code.

I'll create some unittest tomorrow so we can agree on the same results.

ishitasequeira commented:
Also, @woehrl01 just tested your change by making an image and still do see the error.

```
Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 200 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x3661bc0?, 0x71a1d60})
	/go/pkg/mod/k8s.io/apimachinery@v0.26.11/pkg/util/runtime/runtime.go:75 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0005f4da0?})
	/go/pkg/mod/k8s.io/apimachinery@v0.26.11/pkg/util/runtime/runtime.go:49 +0x6b
panic({0x3661bc0?, 0x71a1d60?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
github.com/argoproj/argo-cd/v2/controller/sharding.hasShardingUpdates(...)
	/go/src/github.com/argoproj/argo-cd/controller/sharding/cache.go:154
github.com/argoproj/argo-cd/v2/controller/sharding.(*ClusterSharding).Add(0xc0005f4d80, 0xc00037ad20)
	/go/src/github.com/argoproj/argo-cd/controller/sharding/cache.go:84 +0x131
```

woehrl01 commented:
@ishitasequeira @leoluz if this can wait for a few more hours, I'll add some proper tests for this tomorrow. I'm currently only on my phone which makes this a bit hard.

leoluz commented:
> @ishitasequeira @leoluz if this can wait for a few more hours, I'll add some proper tests for this tomorrow. I'm currently only on my phone which makes this a bit hard.

@woehrl01 Are you able to just revert your last commit? It seems that your first implementation was indeed correct.

woehrl01 commented:
@leoluz done. I also just stumbled over an error in the ismanagedcluster function. Which I fixed as well.

ishitasequeira commented on controller/sharding/cache.go:
```diff
@@ -59,7 +59,7 @@ func (s *ClusterSharding) IsManagedCluster(c *v1alpha1.Cluster) bool {
 	}
 	log.Debugf("Checking if cluster %s with clusterShard %d should be processed by shard %d", c.Server, clusterShard, s.Shard)
-	return clusterShard == s.Shard
+	return s.Shard != nil && clusterShard == int64(*s.Shard)
```
s.Shard is an `int` and so is `clusterShard` here. So no need to change this logic.

woehrl01 replied:
You're right. Reverted, we should rename this values to contain Ptr if they are. That's confusing as hell. 😅

leoluz replied:
I hear you! 😅

ishitasequeira commented:
Okay, so I tested it locally and just changing it `int64(*old.Shard) != int64(*new.Shard)` stops the controller from crashing.

woehrl01 commented:
@ishitasequeira @leoluz 

Thank you for your feedback, I just added some tests. During writing of the tests I discovered some other bugs, and I changed the logic of `hasShardingUpdates` because I now see why it has been confusing before.

The overall problem addressed are:
* The Shard of the "Cluster"- resource should never be modified by the algorithm, it should always be stored in the state. This fixes a bug where round-robin was not deterministic at all, if clusters are added or removed
* The pointer of the cluster was accidentally the pointer of the loop, which in Golang has some strange side effects, I changed that, now the shards are mapped to the correct cluster value (in combination with the previous error this resulted in really strange behaviour in the `Init` function.

I also added some more useful log outputs to see if clusters have changed their shard based on an update/delete event.

Please let me know what you think!

woehrl01 commented:
@leoluz While I was at it, I directly integrated #17183. This makes sense in my opinion because this part was one of the actual needed fixes.

agaudreault commented on controller/sharding/cache.go:
```diff
@@ -136,24 +138,44 @@ func (sharding *ClusterSharding) updateDistribution() {
+		} else {
+			log.Debugf("Cluster %s has not changed shard", k)
+		}
+		sharding.Shards[k] = shard
```
The previous PR seemed to update the cluster shard value in the cache and not only the shard number in the Shards object. 

I think the goal was to persist the computed shard in the cache so we do not need to call `sharding.getClusterShard(c)` and recompute the shard number every time `updateDistribution()` is called. Basically, once the controller starts managing a cluster, it will manage it forever.

Now with the dynamic cluster sharding, I think you are right. We do have to recompute all the cluster shards whenever a cluster is added/modified/deleted.

woehrl01 replied:
@agaudreault Thanks for confirming, the problem with "controller starts managing a cluster" is that this does not work properly as every shard could have a different view on the cluster. If clusters are added/removed and stateful set controllers are restarting at different time intervals (e.g. spot interruptions)

agaudreault reviewed (approved):
LGTM

leoluz reviewed (approved):
LGTM

ishitasequeira reviewed (approved):
LGTM!! Thanks @woehrl01 for working on the fix.

leoluz commented:
/cherry-pick release-2.10

aviveiro commented:
Hey folks, what was the solution here? I reviewed the linked close ticket - was that back-ported to v2.10.0? Should I update `ARGOCD_CONTROLLER_REPLICAS` to `-1`? I am just trying to scale the application controller up to 3 from 1 with the default sharding. Any help here would be appreciated - thank you!
